### PR TITLE
Admin Order - Email is Now Required - Magento 2.3

### PR DESCRIPTION
**Pull request by solwininfotech**
### Fixed Issues (if relevant)

1. magento/magento2#22251: Admin Order - Email is Now Required - Magento 2.2.6 

### Manual testing scenarios (*)

   1. Sales -> Orders
   2. Create New Order
   3. Create New Customer
   4. Select Store View or if Single-Store Mode it brings you to the Order Information Page
   5. Email field is now required

Before Fix:
Customer Email is required in Admin Order Creation.
22251_before
After fixed Customer Email is not required in Admin Order Creation and once order is created customer email will be created automatically by system.
22251_after
22251_dynamic_email_creation

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/magento2?pullRequest=22648) <br/>All committers have signed the CLA.

**Comment by solwininfotech** [comment]
@magento-engcom-team give me test instance

**Comment by magento-engcom-team** [comment]
Hi @solwininfotech. Thank you for your request. I'm working on Magento instance for you

**Comment by magento-engcom-team** [comment]
Hi @solwininfotech, here is your [new Magento instance](https://pr-22648.instances.magento-community.engineering).
Admin access: [https://pr-22648.instances.magento-community.engineering/admin](https://pr-22648.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`

**Comment by rogyar** [comment]
Approving this one.

Basically, this PR reverts the fix provided in https://github.com/magento/magento2/commit/b1eece2f41e99329afd0bb682f2674c0680b258c#diff-b79fbcef70631d52e90840aaf9a0d94d . 

This fix results in confusion since the email auto-generation upon guest order creation from admin panel were considered as a useful feature. And the fix prevents store owners from using the expected behavior. Check the original issue for details.

**Review by rogyar — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @rogyar, thank you for the review.
[ENGCOM-5060](https://jira.corp.magento.com/browse/ENGCOM-5060) has been created to process this Pull Request

**Comment by sdzhepa** [comment]
Hello

@solwininfotech, @trucatchtraps, @rogyar @driskell

Thank you for contribution and collaboration!

Unfortunately, these changes will not be approved by Magento internal team.

Changes from b1eece2 were made by the Core team intentionally.
It was based on several reports and escalations/complaints from different clients through Magento Support channel.
The main idea was that specific of their businesses did not allow to accept any auto-generated email addresses in the case when field is empty(previous behavior).

But after reading all argumentation from issue description #22251
I see that could be different cases based on business needs

So, after several internal discussion, I propose the next solution that should cover both cases:

:exclamation: **NOTE:** 

> Due to several requests, I want to mention that proposed solution and each requirement were 
> discussed and approved with Magento Product Manager @chernenm

_"It should be a configurable option in Configuration"_ with next requirements:

1. The option should be as drop-down element with values Yes/No
2. The option should be *System > Configuration > CUSTOMERS > Customer Configuration* section *Create New Account Options*. Between "Show VAT Number on Storefront" and "Default Email Domain"
3. Element Title: "_Email is required field for Admin order creation_"
4. Hint message under field: "_If set YES Email field will be required during Admin order creation for new Customer._"
5. Element scope: _store view_
6. Default value = _No_
7. The option should affect only Order creation flow for a new customer from Admin side and should not be any impact on other functionality(like Customer creation flow or so)

When option set to "No" Magento should work as it was before b1eece2f41e
When option set to "Yes" Magento should work as now, after

email

**Comment by solwininfotech** [comment]
@sdzhepa  Got your point, we will add the settings in the configuration, thank you for your feedback.

**Comment by solwininfotech** [comment]
@sdzhepa, I have updated the code.
Please view the screens for.
1. Email required setting in the configuration.
email_required_config_setting

2. Create order when Email required setting is No
create_order_when_setting_is_no

3. Auto-generated email once "Create order when Email required setting is No".
auto_generate_email

4. Email required setting is Yes in the configuration.
email_required_config_setting_yes

5. Email required validation for customer email address once "Create order when Email required setting is Yes".
create_order_when_setting_is_yes

6. Admin entered customer email once "Create order when Email required setting is Yes".
admin_entered_email

**Comment by solwininfotech** [comment]
@magento-engcom-team give me test instance

**Comment by magento-engcom-team** [comment]
Hi @solwininfotech. Thank you for your request. I'm working on Magento instance for you

**Comment by magento-engcom-team** [comment]
Hi @solwininfotech, here is your [new Magento instance](https://pr-22648.instances.magento-community.engineering).
Admin access: [https://pr-22648.instances.magento-community.engineering/admin](https://pr-22648.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`

**Comment by solwininfotech** [comment]
@magento-engcom-team 
Testing is done all works fine.

**Comment by sdzhepa** [comment]
@solwininfotech 
Thank you for so quick reaction and code changes!

**Comment by solwininfotech** [comment]
Hi @sdzhepa  @rogyar 

Semantic Version Checker build is filed with minor level because of the following reason.
M071 Constant has been added.
V015 [public] Method has been added.

Can you help me what I need to fix and how or it is okay because of new things are added?
minor_errors

**Comment by solwininfotech** [comment]
@sidolov , Thank you for reviewing this thread let me know I need to update in this.

**Comment by trucatchtraps** [comment]
I finally was able to remove the required * for the email address with a custom module. But now when I try to place an order without an email address, I receive the following errors.

The value of attribute "Email" must be set
Validation is failed.

**Comment by solwininfotech** [comment]
@magento-engcom-team give me test instance

**Comment by magento-engcom-team** [comment]
Hi @solwininfotech. Thank you for your request. I'm working on Magento instance for you

**Comment by magento-engcom-team** [comment]
Hi @solwininfotech, here is your [new Magento instance](https://pr-22648.instances.magento-community.engineering).
Admin access: [https://pr-22648.instances.magento-community.engineering/admin](https://pr-22648.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`

**Comment by solwininfotech** [comment]
Hi @trucatchtraps 

Changes in this branch will be more helpful to you because of this code does not has an error [The value of attribute "Email" must be set Validation is failed.] as well as it has an admin setting that allows you to add an order with or without a customer email address .

[Click here](https://drive.google.com/file/d/1CA5J7ALPtLxCExcX3nm1QLeemy_Nph7Z/view?usp=sharing) for more details.

Thank you.

File: app/code/Magento/Sales/Block/Adminhtml/Order/Create/Form/Account.php [hunk 1]
```diff
@@ -39,6 +39,8 @@ class Account extends AbstractForm
      */
     protected $_extensibleDataObjectConverter;
 
+    const XML_PATH_EMAIL_REQUIRED_CREATE_ORDER = 'customer/create_account/email_required_create_order';
```
Comment (sidolov) [inline app/code/Magento/Sales/Block/Adminhtml/Order/Create/Form/Account.php:42]: Please, make constant private

File: app/code/Magento/Sales/Block/Adminhtml/Order/Create/Form/Account.php [hunk 2]
```diff
@@ -204,4 +206,17 @@ private function extractValuesFromAttributes(array $attributes): array
 
         return $formValues;
     }
+
+    /**
+     * Retrieve email is required field for admin order creation
+     *
+     * @return bool
+     */
+    public function isEmailRequiredCreateOrder()
```
Comment (sidolov) [inline app/code/Magento/Sales/Block/Adminhtml/Order/Create/Form/Account.php:215]: Please, make method private

File: app/code/Magento/Sales/Model/AdminOrder/Create.php [hunk 3]
```diff
@@ -33,6 +33,7 @@ class Create extends \Magento\Framework\DataObject implements \Magento\Checkout\
      */
     const XML_PATH_DEFAULT_EMAIL_DOMAIN = 'customer/create_account/email_domain';
 
+    const XML_PATH_EMAIL_REQUIRED_CREATE_ORDER = 'customer/create_account/email_required_create_order';
```
Comment (sidolov) [inline app/code/Magento/Sales/Model/AdminOrder/Create.php:36]: Please, make constant private

File: app/code/Magento/Sales/Model/AdminOrder/Create.php [hunk 4]
```diff
@@ -2029,7 +2030,30 @@ protected function _validate()
      */
     protected function _getNewCustomerEmail()
     {
-        return $this->getData('account/email');
+        $email_required = $this->_scopeConfig->getValue(
```
Comment (sidolov) [inline app/code/Magento/Sales/Model/AdminOrder/Create.php:2033]: Please, avoid underscore in the variable naming

**Review by sidolov — CHANGES_REQUESTED** [review]

File: app/code/Magento/Sales/Model/AdminOrder/Create.php (hunk 3 above)
Comment (driskell) [reply app/code/Magento/Sales/Model/AdminOrder/Create.php:36]: Judging by the current code not obeying such practice I would suggest that change be made in a separate PR aimed at resolving this in the entire class.

File: app/code/Magento/Sales/Model/AdminOrder/Create.php (hunk 4 above)
Comment (driskell) [reply app/code/Magento/Sales/Model/AdminOrder/Create.php:2033]: As above, this isn’t obeyed by the remaining code and likely require refactor of unrelated code, and so would increase scope of the PR. I’d suggest to pick this up in a separate PR that resolves this for all properties in the class.

File: app/code/Magento/Sales/Model/AdminOrder/Create.php (hunk 3 above)
Comment (driskell) [reply app/code/Magento/Sales/Model/AdminOrder/Create.php:36]: Having said that - I notice the PHPDOC is missing for this constant.

File: app/code/Magento/Sales/Block/Adminhtml/Order/Create/Form/Account.php (hunk 1 above)
Comment (driskell) [reply app/code/Magento/Sales/Block/Adminhtml/Order/Create/Form/Account.php:42]: Needs a PHPDOC too

File: app/code/Magento/Sales/Model/AdminOrder/Create.php (hunk 3 above)
Comment (sidolov) [reply app/code/Magento/Sales/Model/AdminOrder/Create.php:36]: @driskell we can't make existing constants as private due to backward compatibility, but the new one should be private.

File: app/code/Magento/Sales/Model/AdminOrder/Create.php (hunk 4 above)
Comment (sidolov) [reply app/code/Magento/Sales/Model/AdminOrder/Create.php:2033]: @driskell it's not a big deal to rename variables in the modern IDE, it requires an additional commit to make code style better.

**Review by sidolov — CHANGES_REQUESTED** [review]

File: app/code/Magento/Sales/Model/AdminOrder/Create.php (hunk 4 above)
Comment (driskell) [reply app/code/Magento/Sales/Model/AdminOrder/Create.php:2033]: Could it be a problem that it is protected and not private? I worry further change here might delay the PR merge due to unrelated opportunistic change. As it stands I’m happy to help with the other changes but this one feels like it needs a little more thought and a separately considered and reversible changeset so this feature isn’t lost should an issue be raised.

**Comment by solwininfotech** [comment]
Thank you for updated, I am working on this.

**Comment by sidolov** [comment]
@solwininfotech , I am closing this PR now due to inactivity.
Please reopen and update if you wish to continue.
Thank you for the collaboration!

**Comment by trucatchtraps** [comment]
So will these changes be implemented in a new version?

**Comment by solwininfotech** [comment]
Hi @sidolov 
How are you?
Not sure after remote upstream Magento 2 git repo, my this branch was stopped working, I tried all the possible cases but it was not worked so I have done all the changes in a new branch and created a pull request for that. Is that possible you can review it. Here is the branch pull request https://github.com/magento/magento2/pull/24479.

Thank you.

**Pull request closed without merging**